### ISSUE TITLE
question about the reconstruction mechanism

### ISSUE THREAD

**Issue by speedcell4** [minghao-wu/CRF-AE#1]
Hi~
so you just reconstruct the POS tag on the exact position, say P(p_i | enc[w_i, p_i])? instead of predicating the next position POS, say P(p_{i+1} | enc[w_i, p_i])?
I am wondering why this kind of mechanism can help. because, in the exact position, the neural network just know the answer, right? so how can this enhance the performance.

**Comment by minghao-wu** [comment]
Hi there,

This is a great question and thank you.

I believe that the approach you proposed should also work very well. However, designing a over-complicated architecture, which brings minor performance improvement, and staying at the top of the leaderbroad for a few days is not our research problem in this work.

In this work, we are trying to demonstrate the handcrafted features are still of great importance to NLP problems (There are also some works did the same thing as we did). Feature auto-encoder, namely feature reconstruction, is just a tentative attempt to experiment the effectiveness of handcrafted features.

Back to your question, actually, there is no particular reason for why we decided to predict the tag itself, instead of predicting next tag. I just did it. Auto-encoder is also an effective way to extract feature representation, just like language model. Taking the features as input and reconstructing them at the other end, which is exactly how auto-encoder works, can provide more supervision information, such as syntactic structure, which is able to regularize the model resulting in better generalization. 

In addition, the learned contextualized representations by Bi-LSTM does not guarantee the prefect feature prediction, because the input is the concatenation of word vector and feature vectors. The main task and auxiliary tasks share the parameters. 

I believe that having a language-model-like reconstruction architecture should be able to improve the performance further (maybe not). Probably you can run some experiments and share the results with me and others who are interested in.

Feel free to ask, if you have any further question.

**Comment by speedcell4** [comment]
Thank you for your comprehensively reply.

another question is about how to reduce the loss score in a batch.
for NER prediction, you utilized the CRF, but, to my understanding, you just [sum them up](https://github.com/minghao-wu/CRF-AE/blob/master/model.py#L148)
then for POS reconstruction, you used the [averaged cross entropy](https://github.com/minghao-wu/CRF-AE/blob/master/model.py#L202). (the default value of `reduction` argument for `cross_entropy`function  is `"mean"`)
so is this the reason to set lambda just `1`, since POS reconstruction loss is small enough?

so would `sum-NER-loss + other-lambda * sum-POS-loss` performs better? 
by the way, actually I did not find where you set the `batch_size`, could you please show me the exact place?

**Comment by minghao-wu** [comment]
The baseline model is proposed by Xuezhe Ma and codebase is adapted from [Zhixiu Ye's implementation](https://github.com/ZhixiuYe/NER-pytorch) with minor modification.

For the main task, NER, batch_size = 1, and hence, there is no need to average this one.
For the auxiliary tasks, feature reconstruction, batch_size = sequence length.

To be honest, the choice of lambda is one of the weakness for this work, since you have to spend plenty of efforts on different choices of lambdas. If the lambda is too large or too small, the auxiliary tasks could overwhelm the main one or be unable to provide enough impacts. 

There is also no particular reason for why I set lambda to be 1. Experiments talk. I was very lucky to find the optimal lambda in my first intuitive attempt.
